File: checkers/checkers.py
""" Defines the checkers board and implements the rules of the game.
"""
import hashlib


class CheckersBoard():
    """ CheckersBoard class and associated methods.
    """

    WHITE = 1
    BLACK = 2
    WHITE_KING = 3
    BLACK_KING = 4

    BOARD_SIZE = 8
    NUM_PIECES = 12  # Number of pieces per player

    def __init__(self, initial_state=None):
        if initial_state is None:
            self.board = CheckersBoard.initial_board()
        else:
            self.board = [x[:] for x in initial_state]
        self._hash_value = None

    def __eq__(self, other):
        return self.board == other.board

    def __hash__(self):
        if self._hash_value is not None:
            return self._hash_value

        pieces = self.pieces()
        h = hashlib.md5()
        for piece in pieces:
            h.update(bytes(piece))
        v = int(h.hexdigest()[-15:], 16)

        self._hash_value = v
        return v

    @classmethod
    def initial_board(cls):
        board = [[0] * cls.BOARD_SIZE for _ in range(cls.BOARD_SIZE)]
        for i in range(3):
            offset = 0 if i == 1 else 1
            for j in range(4):
                n = j * 2 + offset
                board[i][n] = cls.WHITE

        for i in range(3):
            offset = 1 if i == 1 else 0
            for j in range(4):
                n = j * 2 + offset
                board[7 - i][n] = cls.BLACK

        return board

    @classmethod
    def copy(cls, obj):
        return CheckersBoard(initial_state=obj.board)

    def count(self):
        """ Returns the number of white and black pieces in the board.
        """
        w_count = 0
        b_count = 0
        for i in range(self.BOARD_SIZE):
            for j in range(self.BOARD_SIZE):
                v = self.board[i][j]
                if v in [self.WHITE, self.WHITE_KING]:
                    w_count += 1
                elif v in [self.BLACK, self.BLACK_KING]:
                    b_count += 1

        return w_count, b_count

    def piece_count(self):
        """ Returns the per piece type counts.
        """
        counts = [0] * 4
        for i in range(self.BOARD_SIZE):
            for j in range(self.BOARD_SIZE):
                v = self.board[i][j]
                if v == 0:
                    continue
                counts[v - 1] += 1
        return counts

    @classmethod
    def _is_player_piece(cls, v, player):
        if player == cls.WHITE:
            return v in [cls.WHITE, cls.WHITE_KING]
        elif player == cls.BLACK:
            return v in [cls.BLACK, cls.BLACK_KING]
        return False

    @classmethod
    def _is_opposing_player_piece(cls, v, player):
        if player == cls.WHITE:
            return v in [cls.BLACK, cls.BLACK_KING]
        elif player == cls.BLACK:
            return v in [cls.WHITE, cls.WHITE_KING]
        return False

    def _generate_piece_moves(self, row, col, player, capture_only=False):
        """ pieces (men) can move forwards only.
        """
        nrow = row + 1 if player == self.WHITE else row - 1
        if nrow < 0 or nrow >= self.BOARD_SIZE:
            return [], []

        capture_moves = []
        non_capture_moves = []

        def gen_capture_move(nnrow, nncol):
            additional, _ = self._generate_piece_moves(
                nnrow, nncol, player, capture_only=True)
            if additional:
                for m in additional:
                    capture_moves.append([(row, col)] + m)
            else:
                capture_moves.append([(row, col), (nnrow, nncol)])

        # left diagnonal
        if col > 0:
            v = self.board[nrow][col-1]
            if v == 0:
                if not capture_only:
                    non_capture_moves.append([(row, col), (nrow, col - 1)])
            elif col > 1 and self._is_opposing_player_piece(v, player):
                nnrow = nrow + 1 if player == self.WHITE else nrow - 1
                if nnrow >= 0 and nnrow < 8 and self.board[nnrow][col - 2] == 0:
                    gen_capture_move(nnrow, col - 2)

        # right diagnonal
        if col < self.BOARD_SIZE - 1:
            v = self.board[nrow][col+1]
            if v == 0:
                if not capture_only:
                    non_capture_moves.append([(row, col), (nrow, col + 1)])
            elif (col < self.BOARD_SIZE - 2 and
                  self._is_opposing_player_piece(v, player)):
                nnrow = nrow + 1 if player == self.WHITE else nrow - 1
                if (nnrow >= 0 and nnrow < self.BOARD_SIZE and
                        self.board[nnrow][col + 2] == 0):
                    gen_capture_move(nnrow, col + 2)

        return capture_moves, non_capture_moves

    def _generate_king_moves(self, row, col, player, capture_only=False,
                             exclude_dir=None, exclude_list=None, depth=0):
        """ kings can move in any diagonal (forward or backwards) and capture
            an opposing piece if there is an empty space beyond it.
        """
        def _advance(pos, vdir, hdir):
            row = pos[0] + (1 if vdir else -1)
            col = pos[1] + (1 if hdir else -1)
            return row, col

        def _is_valid_position(pos):
            return all(c >= 0 and c < self.BOARD_SIZE for c in pos)

        capture_moves = []
        non_capture_moves = []

        start = (row, col)
        if exclude_list is None:
            exclude_list = []

        for diagonal in range(4):
            vdir = diagonal // 2
            hdir = diagonal % 2

            if (vdir, hdir) == exclude_dir:
                continue
            pos = start
            capture = False
            while True:
                pos = _advance(pos, vdir, hdir)
                if not _is_valid_position(pos):
                    break
                v = self.get_position(pos)
                if self._is_player_piece(v, player):
                    break
                if pos in exclude_list:
                    break
                if self._is_opposing_player_piece(v, player):
                    npos = _advance(pos, vdir, hdir)
                    if not _is_valid_position(npos):
                        break
                    if self.get_position(npos) != 0:
                        break
                    capture = True
                    exclude_list.append(pos)
                    pos = npos

                if capture:
                    assert depth < 32
                    additional, _ = self._generate_king_moves(
                        pos[0], pos[1], player, capture_only=True,
                        exclude_dir=(not vdir, not hdir),
                        exclude_list=exclude_list.copy(),
                        depth=depth + 1)
                    if additional:
                        capture_moves.extend([[start] + m for m in additional])
                        continue

                if capture:
                    capture_moves.append([start, pos])
                elif not capture_only:
                    non_capture_moves.append([start, pos])

        return capture_moves, non_capture_moves

    def valid_moves(self, player, select_capture=False):
        """
          Returns: list of valid moves where a move is described by a tuple
          with begin, end coordinates for a piece.
        """
        if player not in [self.BLACK, self.WHITE]:
            raise ValueError('Invalid value for player: %r', player)

        move_lists = [[], []]
        for i in range(8):
            for j in range(8):
                v = self.board[i][j]
                if not self._is_player_piece(v, player):
                    continue

                if player == v:
                    pos_moves = self._generate_piece_moves(i, j, player)
                else:
                    pos_moves = self._generate_king_moves(i, j, player)

                for gbl_list, pos_list in zip(move_lists, pos_moves):
                    assert all(self.get_position(pos) == 0
                               for move in pos_list for pos in move[1:])
                    gbl_list.extend(pos_list)

        if select_capture:
            if move_lists[0]:
                return move_lists[0]
            return move_lists[1]
        return move_lists[0] + move_lists[1]

    def valid_position_moves(self, position, select_capture=False):
        row, col = position
        v = self.board[row][col]
        if v in [self.WHITE, self.WHITE_KING]:
            player = self.WHITE
        elif v in [self.BLACK, self.BLACK_KING]:
            player = self.BLACK
        else:
            raise ValueError('No piece in specified position')

        if v in [self.WHITE, self.BLACK]:
            capture, non_capture = self._generate_piece_moves(row, col, player)
        if v in [self.WHITE_KING, self.BLACK_KING]:
            capture, non_capture = self._generate_king_moves(row, col, player)

        if select_capture and capture:
            moves = capture
        else:
            moves = capture + non_capture
        assert all(self.get_position(pos) == 0
                   for move in moves for pos in move[1:])
        return moves

    def get_position(self, coordinates):
        return self.board[coordinates[0]][coordinates[1]]

    def _clear_position(self, coordinates):
        self.board[coordinates[0]][coordinates[1]] = 0

    def _set_position(self, coordinates, value):
        self.board[coordinates[0]][coordinates[1]] = value

    def _player_count(self, player):
        """ Check how many pieces this player has. If the count is less than
            12, the player has a piece that has been captured and can be used
            to crown a king.
        """
        count = 0
        for i in range(8):
            for j in range(8):
                v = self.board[i][j]
                if v == player:
                    count += 1
                if ((player == self.WHITE and v == self.WHITE_KING) or
                        (player == self.BLACK and v == self.BLACK_KING)):
                    count += 2
        return count

    def _maybe_promote_piece(self, coordinates, piece, player):
        if piece not in [self.WHITE, self.BLACK]:
            return piece
        if player == self.WHITE and coordinates[0] == 7:
            return self.WHITE_KING
        elif player == self.BLACK and coordinates[0] == 0:
            return self.BLACK_KING
        return piece

    def move(self, move, enable_validation=False):
        """ Changes the board by moving a piece.
            It verifies whether the move is valid and will eat pieces.

            Arguments:
              move: A list of 2 or more positions. Simple moves have only
              a start and end position; however it is possible to eat
              multiple pieces in one move. In this case, the move includes
              the intermediate positions since they may include changes in
              direction.
        """
        start = move[0]
        piece = self.get_position(start)
        if piece in [self.WHITE, self.WHITE_KING]:
            player = self.WHITE
        elif piece in [self.BLACK, self.BLACK_KING]:
            player = self.BLACK
        else:
            raise ValueError(f'Invalid start position {start}')

        if enable_validation:
            allowed_moves = self.valid_position_moves(start)
            if move not in allowed_moves:
                raise ValueError(f'Invalid move {move}')

        self._hash_value = None
        self._clear_position(start)
        pos = start

        def _incr(dir):
            return 1 if dir else - 1

        for i in range(1, len(move)):
            npos = move[i]
            assert self.get_position(npos) == 0
            vdir = npos[0] > pos[0]
            hdir = npos[1] > pos[1]

            while True:
                pos = (pos[0] + _incr(vdir), pos[1] + _incr(hdir))
                if pos == npos:
                    break
                v = self.get_position(pos)
                if v != 0:
                    assert self._is_opposing_player_piece(v, player)
                    self._clear_position(pos)

        piece = self._maybe_promote_piece(move[-1], piece, player)
        self._set_position(move[-1], piece)

    def pieces(self):
        """ Returns the list of all board pieces as (row, col, piece code)
        """
        results = []
        for i in range(self.BOARD_SIZE):
            for j in range(self.BOARD_SIZE):
                piece = self.board[i][j]
                if piece == 0:
                    continue
                results.append((i, j, piece))
        return results

File: checkers/checkers.cc
#include "checkers.h"

#include <functional>
#include <iostream>
#include <sstream>

using namespace std;

pair<CheckersBoard::Position, bool> position_advance(
        CheckersBoard::Position pos, bool vdir, bool hdir) {
    int row = pos.first + (vdir ? 1 : -1);
    int col = pos.second + (hdir ? 1 : -1);
    if (row >= 0 && row < CheckersBoard::BOARD_SIZE && col >= 0 &&
        col < CheckersBoard::BOARD_SIZE) {
        return make_pair(make_pair(row, col), true);
    }
    return make_pair(make_pair(-1, -1), false);
}

bool is_player_piece(CheckersBoard::Piece piece, CheckersBoard::Player player) {
    switch (piece) {
        case CheckersBoard::Piece::WHITE_PAWN:
        case CheckersBoard::Piece::WHITE_KING:
            return player == CheckersBoard::Player::WHITE;
        case CheckersBoard::Piece::BLACK_PAWN:
        case CheckersBoard::Piece::BLACK_KING:
            return player == CheckersBoard::Player::BLACK;
    }
    return false;
}

bool is_opposing_player_piece(CheckersBoard::Piece piece,
                              CheckersBoard::Player player) {
    return piece != 0 && !is_player_piece(piece, player);
}

namespace {

ostream &operator<<(ostream &os, const CheckersBoard::Position &pos) {
    os << "[" << pos.first << ", " << pos.second << "]";
    return os;
}

ostream &operator<<(ostream &os, const CheckersBoard::Move &move) {
    os << "[";
    for (int i = 0; i < move.size(); i++) {
        if (i > 0) {
            os << ", ";
        }
        os << move[i];
    }
    os << "]";
    return os;
}

template <typename C>
std::string repr(const C &v) {
    stringstream ss;
    ss << v;
    return ss.str();
}

}  // namespace

CheckersBoard::CheckersBoard() {
    board_.fill(0);
    for (int i = 0; i < 3; i++) {
        int offset = (i == 1) ? 0 : 1;
        for (int j = 0; j < 4; j++) {
            int col = j * 2 + offset;
            board_[i * BOARD_SIZE + col] = Piece::WHITE_PAWN;
        }
    }

    for (int i = 0; i < 3; i++) {
        int offset = (i == 1) ? 1 : 0;
        int row = BOARD_SIZE - i - 1;
        for (int j = 0; j < 4; j++) {
            int col = j * 2 + offset;
            board_[row * BOARD_SIZE + col] = Piece::BLACK_PAWN;
        }
    }
}
CheckersBoard::CheckersBoard(const CheckersBoard &rhs) {
    memcpy(board_.data(), rhs.board_.data(), board_.size());
}

CheckersBoard::CheckersBoard(const std::vector<int> &state) {
    if (state.size() != BOARD_SIZE * BOARD_SIZE) {
        throw std::invalid_argument("argument size");
    }
    for (int i = 0; i < board_.size(); i++) {
        board_[i] = state[i];
    }
}

void CheckersBoard::initialize(const std::vector<int> &state) {
    if (state.size() != BOARD_SIZE * BOARD_SIZE) {
        throw std::invalid_argument("argument size");
    }
    for (int i = 0; i < board_.size(); i++) {
        board_[i] = state[i];
    }
}

CheckersBoard::Piece CheckersBoard::maybe_promote_piece(Position position,
                                                        Piece piece,
                                                        Player player) const {
    if (player == Player::WHITE && position.first == BOARD_SIZE - 1) {
        return Piece::WHITE_KING;
    }
    if (player == Player::BLACK && position.first == 0) {
        return Piece::BLACK_KING;
    }
    return piece;
}

CheckersBoard::Piece CheckersBoard::get_position(Position pos) const {
    return static_cast<CheckersBoard::Piece>(
            board_[pos.first * BOARD_SIZE + pos.second]);
}

void CheckersBoard::clear_position(Position pos) {
    board_[pos.first * BOARD_SIZE + pos.second] = 0;
}

void CheckersBoard::set_position(Position pos, Piece piece) {
    board_[pos.first * BOARD_SIZE + pos.second] = piece;
}

std::pair<std::list<CheckersBoard::Move>, std::list<CheckersBoard::Move>>
CheckersBoard::generate_pawn_moves(Position start, Player player,
                                   bool capture_only,
                                   const std::list<Position> &prefix) const {
    bool vdir = player == Player::WHITE ? true : false;

    std::list<Move> capture_moves;
    std::list<Move> non_capture_moves;

    for (int i = 0; i < 2; i++) {
        bool hdir = i > 0;

        Position pos;
        bool ok;
        tie(pos, ok) = position_advance(start, vdir, hdir);
        if (!ok) {
            continue;
        }
        Piece piece = get_position(pos);
        if (piece == 0) {
            if (!capture_only) {
                Move move;
                move.reserve(prefix.size() + 2);
                copy(prefix.begin(), prefix.end(), std::back_inserter(move));
                move.push_back(start);
                move.push_back(pos);
                non_capture_moves.push_back(std::move(move));
            }
            continue;
        }
        if (!is_opposing_player_piece(piece, player)) {
            continue;
        }
        Position npos;
        tie(npos, ok) = position_advance(pos, vdir, hdir);
        if (!ok || get_position(npos) != 0) {
            continue;
        }

        std::list<Move> sub_capture;
        std::list<Move> sub_non_capture;

        list<Position> nprefix(prefix);
        nprefix.push_back(start);

        tie(sub_capture, sub_non_capture) =
                generate_pawn_moves(npos, player, true, nprefix);
        if (!sub_capture.empty()) {
            capture_moves.splice(capture_moves.end(), std::move(sub_capture));
            continue;
        }

        Move move;
        move.reserve(prefix.size() + 2);
        copy(prefix.begin(), prefix.end(), std::back_inserter(move));
        move.push_back(start);
        move.push_back(npos);
        capture_moves.push_back(std::move(move));
    }

    return make_pair(capture_moves, non_capture_moves);
}

std::pair<std::list<CheckersBoard::Move>, std::list<CheckersBoard::Move>>
CheckersBoard::generate_king_moves(Position start, Player player,
                                   bool recursion,
                                   std::pair<bool, bool> exclude_dir,
                                   const std::list<Position> &exclude_list,
                                   const std::list<Position> &prefix) const {
    std::list<Move> capture_moves;
    std::list<Move> non_capture_moves;

    for (int diagonal = 0; diagonal < 4; diagonal++) {
        bool vdir = diagonal >> 1;
        bool hdir = diagonal & 1;

        if (recursion && make_pair(vdir, hdir) == exclude_dir) {
            continue;
        }

        Position pos = start;
        bool capture = false;

        std::list<Move> non_recursive_captures;

        list<Position> n_exclude_list(exclude_list);
        while (true) {
            bool ok;
            tie(pos, ok) = position_advance(pos, vdir, hdir);
            if (!ok) {
                break;
            }
            Piece piece = get_position(pos);
            if (is_player_piece(piece, player)) {
                break;
            }
            if (std::find(exclude_list.begin(), exclude_list.end(), pos) !=
                exclude_list.end()) {
                break;
            }

            if (is_opposing_player_piece(piece, player)) {
                if (capture) {
                    break;
                }
                Position npos;
                bool ok;
                tie(npos, ok) = position_advance(pos, vdir, hdir);
                if (!ok || get_position(npos) != 0) {
                    break;
                }
                capture = true;
                n_exclude_list.push_back(pos);
                pos = npos;
            }

            if (capture) {
                std::list<Move> sub_capture;
                std::list<Move> sub_non_capture;
                list<Position> nprefix(prefix);
                nprefix.push_back(start);
                tie(sub_capture, sub_non_capture) = generate_king_moves(
                        pos, player, true, make_pair(!vdir, !hdir),
                        n_exclude_list, nprefix);
                if (!sub_capture.empty()) {
                    capture_moves.splice(capture_moves.end(),
                                         std::move(sub_capture));
                    non_recursive_captures.clear();
                    break;
                }
            }

            Move move;
            move.reserve(prefix.size() + 2);
            copy(prefix.begin(), prefix.end(), std::back_inserter(move));
            move.push_back(start);
            move.push_back(pos);

            if (capture) {
                non_recursive_captures.push_back(std::move(move));
            } else if (!recursion) {
                non_capture_moves.push_back(std::move(move));
            }
        }
        capture_moves.splice(capture_moves.end(),
                             std::move(non_recursive_captures));
    }

    return make_pair(capture_moves, non_capture_moves);
}

bool CheckersBoard::operator==(const CheckersBoard &rhs) const {
    return memcmp(board_.data(), rhs.board_.data(), board_.size()) == 0;
}

uint64_t CheckersBoard::hash() const {
    return std::hash<string>{}(string(board_.data(), board_.size()));
}

// Returns the number of pieces for each player.
std::pair<int, int> CheckersBoard::count() const {
    int w_count = 0;
    int b_count = 0;

    for (int i = 0; i < BOARD_SIZE * BOARD_SIZE; i++) {
        switch (board_[i]) {
            case CheckersBoard::Piece::WHITE_PAWN:
            case CheckersBoard::Piece::WHITE_KING:
                w_count++;
                break;
            case CheckersBoard::Piece::BLACK_PAWN:
            case CheckersBoard::Piece::BLACK_KING:
                b_count++;
                break;
        }
    }
    return make_pair(w_count, b_count);
}

// Returns the per piece type counts.
std::array<int, 4> CheckersBoard::piece_count() const {
    std::array<int, 4> counts;
    counts.fill(0);

    for (int i = 0; i < BOARD_SIZE * BOARD_SIZE; i++) {
        int piece = board_[i];
        if (piece != 0) {
            counts[piece - 1] += 1;
        }
    }
    return counts;
}

// Returns the lists of possible capture and non-capture moves for
// the specified player.
std::list<CheckersBoard::Move> CheckersBoard::valid_moves(Player player) const {
    std::list<Move> capture_moves;
    std::list<Move> non_capture_moves;

    for (int i = 0; i < BOARD_SIZE * BOARD_SIZE; i++) {
        Piece piece = static_cast<Piece>(board_[i]);
        if (!is_player_piece(piece, player)) {
            continue;
        }
        Position pos = {i >> 3, i & 7};

        list<Move> pos_capture;
        list<Move> pos_non_capture;
        if (piece == Piece::WHITE_KING || piece == Piece::BLACK_KING) {
            tie(pos_capture, pos_non_capture) =
                    generate_king_moves(pos, player, false, {}, {}, {});
        } else {
            tie(pos_capture, pos_non_capture) =
                    generate_pawn_moves(pos, player, false, {});
        }
        capture_moves.splice(capture_moves.end(), std::move(pos_capture));
        non_capture_moves.splice(non_capture_moves.end(),
                                 std::move(pos_non_capture));
    }

    // Implement the rule that player must capture pieces picking the highest
    // capture length.
    if (!capture_moves.empty()) {
        size_t maxlen = 0;
        for (auto &m : capture_moves) {
            maxlen = std::max(maxlen, m.size());
        }
        std::list<Move> max_sz_captures;
        std::copy_if(capture_moves.begin(), capture_moves.end(),
                     std::back_inserter(max_sz_captures),
                     [maxlen](const Move &m) { return m.size() == maxlen; });
        return max_sz_captures;
    }

    return non_capture_moves;
}

// Returns the lists of possible capture and non-capture moves for the
// piece in the specified position.
std::pair<std::list<CheckersBoard::Move>, std::list<CheckersBoard::Move>>
CheckersBoard::position_moves(Position pos) const {
    std::list<CheckersBoard::Move> capture_moves;
    std::list<CheckersBoard::Move> non_capture_moves;

    Piece piece = get_position(pos);
    switch (piece) {
        case WHITE_PAWN:
            tie(capture_moves, non_capture_moves) =
                    generate_pawn_moves(pos, Player::WHITE, false, {});
            break;
        case BLACK_PAWN:
            tie(capture_moves, non_capture_moves) =
                    generate_pawn_moves(pos, Player::BLACK, false, {});
            break;
        case WHITE_KING:
            tie(capture_moves, non_capture_moves) =
                    generate_king_moves(pos, Player::WHITE, false, {}, {}, {});
            break;
        case BLACK_KING:
            tie(capture_moves, non_capture_moves) =
                    generate_king_moves(pos, Player::BLACK, false, {}, {}, {});
            break;
    }
    return make_pair(capture_moves, non_capture_moves);
}

std::list<CheckersBoard::PositionPiece> CheckersBoard::pieces() const {
    std::list<CheckersBoard::PositionPiece> pieces;
    for (int i = 0; i < board_.size(); i++) {
        Piece piece = static_cast<Piece>(board_[i]);
        if (piece == 0) {
            continue;
        }
        int row = i >> 3;
        int col = i & 7;
        pieces.push_back(PositionPiece({static_cast<uint8_t>(row),
                                        static_cast<uint8_t>(col),
                                        static_cast<uint8_t>(piece)}));
    }
    return pieces;
}

void CheckersBoard::move(const CheckersBoard::Move &move) {
    Position start = move[0];
    Piece piece = get_position(start);
    Player player;

    switch (piece) {
        case CheckersBoard::Piece::WHITE_PAWN:
        case CheckersBoard::Piece::WHITE_KING:
            player = CheckersBoard::Player::WHITE;
            break;
        case CheckersBoard::Piece::BLACK_PAWN:
        case CheckersBoard::Piece::BLACK_KING:
            player = CheckersBoard::Player::BLACK;
            break;
        default:
            throw std::invalid_argument("Invalid start position " +
                                        repr(start));
    }

    clear_position(start);
    Position pos = start;

    for (int i = 1; i < move.size(); i++) {
        Position npos = move[i];
        if (get_position(npos) != 0) {
            throw std::invalid_argument("Invalid move position " + repr(npos));
        }
        bool vdir = npos.first > pos.first;
        bool hdir = npos.second > pos.second;

        while (true) {
            bool ok;
            tie(pos, ok) = position_advance(pos, vdir, hdir);
            if (!ok) {
                throw std::out_of_range("Invalid position " + repr(pos));
            }
            if (pos == npos) {
                break;
            }
            Piece intermediate = get_position(pos);
            if (intermediate != 0) {
                if (!is_opposing_player_piece(intermediate, player)) {
                    throw std::invalid_argument("Invalid capture " + repr(pos));
                }
                clear_position(pos);
            }
        }
    }

    Position end = move[move.size() - 1];
    piece = maybe_promote_piece(end, piece, player);
    set_position(end, piece);
}
